build/cmake/GenerateUninstaller.cmake

```diff
@@ -0,0 +1,271 @@
+cmake_policy(VERSION 3.15)
+
+if(NOT CMAKE_SCRIPT_MODE_FILE)
+    # We are being included from within a project, so we should generate the install rules
+    # The script name is "uninstall" by default:
+    if(NOT DEFINED UNINSTALL_SCRIPT_NAME)
+        set(UNINSTALL_SCRIPT_NAME "uninstall")
+    endif()
+    # We need a directory where we should install the script:
+    if(NOT UNINSTALL_PROG_DIR)
+        message(SEND_ERROR "We require an UNINSTALL_PROG_DIR to be defined")
+    endif()
+    # Platform dependent values:
+    if(WIN32)
+        set(_script_ext "cmd")
+        set(_script_runner cmd.exe /c)
+    else()
+        set(_script_ext "sh")
+        set(_script_runner sh -e -u)
+    endif()
+    # The script filename and path:
+    set(_script_filename "${UNINSTALL_SCRIPT_NAME}.${_script_ext}")
+    get_filename_component(_uninstaller_script "${CMAKE_CURRENT_BINARY_DIR}/${_script_filename}" ABSOLUTE)
+    # Code that will do the work at install-time:
+    string(CONFIGURE [==[
+    function(__generate_uninstall)
+        set(UNINSTALL_IS_WIN32 "@WIN32@")
+        set(UNINSTALL_WRITE_FILE "@_uninstaller_script@")
+        set(UNINSTALL_SCRIPT_SELF "@UNINSTALL_PROG_DIR@/@_script_filename@")
+        include("@CMAKE_CURRENT_LIST_FILE@")
+    endfunction()
+    __generate_uninstall()
+    ]==] code @ONLY ESCAPE_QUOTES)
+    install(CODE "${code}")
+    # Add a rule to install that file:
+    install(
+        FILES "${_uninstaller_script}"
+        DESTINATION "${UNINSTALL_PROG_DIR}"
+        PERMISSIONS
+            OWNER_READ OWNER_WRITE OWNER_EXECUTE
+            GROUP_READ GROUP_EXECUTE
+            WORLD_READ WORLD_EXECUTE
+        )
+
+    # If applicable, generate an "uninstall" target to run the uninstaller:
+    if(CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR OR PROJECT_IS_TOP_LEVEL)
```

Comment:
Is this how a collision is prevented between the uninstall generated in this project and, for instance, the uninstall generated in the C++ driver when it includes the C driver directly?

Reply:
Yes. This is mostly the same as what was done previously, but with the addition of the newer CMake feature `PROJECT_IS_TOP_LEVEL`, which was added to address this exact use case (on older CMake, that variable check will be a no-op).